File: 109. Convert Sorted List to Binary Search Tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        if not head:
            return head
        ls = []
        while head:
            ls.append(head.val)
            head = head.next
        return self.helper(ls, 0, len(ls)-1)
        
    def helper(self, nums, lo, hi):
        if lo > hi or lo < 0 or hi > len(nums)-1:
            return 
        elif lo == hi:
            return TreeNode(nums[lo])
        else:
            mid = (hi - lo) // 2 + lo
            node = TreeNode(nums[mid])
            node.left = self.helper(nums, lo, mid-1)
            node.right = self.helper(nums, mid+1, hi)
        return node
            

File: 109. Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private List<Integer> array;
    
    public Solution() {
        this.array = new ArrayList<Integer>();
    }
    
    public TreeNode sortedListToBST(ListNode head) {
        while (head != null) {
            this.array.add(head.val);
            head = head.next;
        }
        TreeNode root = ListToBST(0, array.size()-1);
        return root;
    }
    
    private TreeNode ListToBST(int left, int right) {
        if (left > right) {
            return null;
        } else if (left == right) {
            return new TreeNode(this.array.get(left));
        } else {
            int mid = (right-left) / 2 + left;
            TreeNode node = new TreeNode(this.array.get(mid));
            node.left = ListToBST(left, mid-1);
            node.right = ListToBST(mid+1, right);
            return node;
        }
    }
}